blog/2024-09-12-building-the-opentofu-registry.md

```diff
@@ -0,0 +1,129 @@
+
+Here we owe a big thanks to Cloudflare. Their very competitive pricing for [R2](https://www.cloudflare.com/developer-platform/r2/) with no egress fees and their sponsorship of OpenTofu meant that we would be able to run the Registry essentially for free with no servers and scaling issues to worry about. The [Registry codebase](https://github.com/opentofu/registry) (written in Go) pre-generates all possible answers of the API above and uploads the static files to an R2 bucket.
+
+## Populating the Registry
+
+Alongside the Terraform license change, HashiCorp [closed the Terraform Registry](https://registry.terraform.io/terms) for software that isn't Terraform, so using it as a source of data for the OpenTofu Registry was out of the question. However, since the Terraform Registry was coupled exclusively to GitHub, it was relatively straight forward for us to use the GitHub search API to populate the Registry, if a little slow.
+
+Updating the Registry, however, was a much harder problem. GitHub limits the API to ~5,000 requests per hour, which is not enough to update roughly 30,000 providers and modules in a speedy fashion, especially since some updates required multiple requests.
+
+We *could* have requested provider authors to log in with their GitHub account, granting us an access token we could use for a higher rate limit, but that would have been impractical when we launched the Registry since we would have had to ask thousands of developers to log in at a time when OpenTofu wasn't even released yet.
+
+The solution came by the way of the GitHub RSS feeds, which are not rate limited. The releases RSS located at `http://github.com/USERNAME/REPO/releases.atom` always contains the last 5 releases, which is sufficient if we only need to add the latest versions as it is unlikely that a provider or module would have more than 5 releases within an hour. For modules, we needed to query the tags instead of the releases, which was even easier because the `git ls-remote` command gave us all this information and was also not rate limited. (You are going to keep it that way, right, GitHub?)
+
+### The submission process
+
+Since we didn't need to ask provider and module authors for their credentials using OAuth, we were able to create a simple submission process that anyone with a GitHub account could use. 
+
+While anyone could submit a provider or a module, we still needed provider authors to submit their GPG keys so their binaries could be verified. Instead of asking for OAuth logins, we again decided to use the GitHub API in a creative way. In order to submit a GPG key for a provider, the author needed to set their organization membership to public for the repository of the provider and then submit a GitHub issue. We would verify their membership in the provider organization and process their GPG key.
```

Comment:
```suggestion
While anyone could submit a provider or a module, we still needed provider authors to submit their GPG keys so their binaries could be verified. Instead of asking for OAuth logins, we again decided to use the GitHub API in a creative way. In order to submit a GPG key for a provider, the author needed to set their organization membership to public for the repository of the provider and then submit a GitHub issue. We would verify their membership in the provider organization and process their GPG keys.
```